File: commons-parsers/src/main/java/xyz/wagyourtail/commons/parsers/impl/json5/Json5Array.java
package xyz.wagyourtail.commons.parsers.impl.json5;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.val;
import xyz.wagyourtail.commons.core.CollectionUtils;
import xyz.wagyourtail.commons.core.reader.CharReader;
import xyz.wagyourtail.commons.core.reader.StringCharReader;
import xyz.wagyourtail.commons.parsers.Data;
import xyz.wagyourtail.commons.parsers.StringData;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Json5Array extends StringData.OnlyParsed<Json5Array.ArrayContent> {


    public Json5Array(ArrayContent content) {
        super(content);
    }

    public static Json5Array parse(String raw) {
        StringCharReader charReader = new StringCharReader(raw);
        val content = contentBuilder(charReader);
        charReader.expectEOS();
        return new Json5Array(content);
    }

    public static Json5Array parse(CharReader<?> charReader) {
        return new Json5Array(contentBuilder(charReader));
    }

    public static ArrayContent contentBuilder(CharReader<?> reader) {
        List<Object> entries = new ArrayList<>();
        List<Json5Value> entryList = new ArrayList<>();

        reader.expect('[');
        while (!reader.exhausted()) {
            while (!reader.exhausted()) {
                val value = reader.parseOrNull(
                        Json5Whitespace::new,
                        Json5Comment::new
                );
                if (value == null) break;
                entries.add(value);
            }
            if (reader.peek() == ']') {
                break;
            }

            val entry = Json5Value.parse(reader);
            entries.add(entry);
            entryList.add(entry);

            while (!reader.exhausted()) {
                val value = reader.parseOrNull(
                        Json5Whitespace::new,
                        Json5Comment::new
                );
                if (value == null) break;
                entries.add(value);
            }
            if (reader.peek() == ',') {
                entries.add((char) reader.take());
            } else {
                break;
            }
        }
        reader.expect(']');
        return new ArrayContent(entries, entryList);
    }

    @Getter
    @AllArgsConstructor
    public static class ArrayContent extends Data.Content {
        private final List<Object> entries;
        private final List<Json5Value> values;

        public ArrayContent(List<Object> entries) {
            this.entries = entries;
            this.values = new ArrayList<>();
            for (Object entry : entries) {
                if (entry instanceof Json5Value) {
                    values.add((Json5Value) entry);
                }
            }
        }

        @Override
        public Iterable<Object> getEntries() {
            return CollectionUtils.concat(
                    Collections.singleton('['),
                    entries,
                    Collections.singleton(']')
            );
        }
    }

}

File: commons-parsers-kt-kt/src/commonMain/kotlin/xyz/wagyourtail/commonskt/parsers/impl/json5/Json5Array.kt
package xyz.wagyourtail.commonskt.parsers.impl.json5

import xyz.wagyourtail.commonskt.parsers.StringData
import xyz.wagyourtail.commonskt.reader.CharReader
import xyz.wagyourtail.commonskt.utils.iterable

class Json5Array(content: ArrayContent) : StringData.OnlyParsed<Json5Array.ArrayContent>(content) {

    companion object : StringDataBuilder<Json5Array, ArrayContent> {

        override fun invoke(rawContent: CharReader<*>): Json5Array {
            return Json5Array(checkedBuildContent(rawContent))
        }

        override fun checkedBuildContent(reader: CharReader<*>): ArrayContent {
            val entries = mutableListOf<Any>()
            val values = mutableListOf<Json5Value>()
            reader.expect('[')
            while (!reader.exhausted()) {
                while (!reader.exhausted()) {
                    val value = reader.parseOrNull(
                        Json5Whitespace::invoke,
                        Json5Comment::invoke
                    )
                    if (value == null) break
                    entries.add(value)
                }
                if (reader.peek() == ']') {
                    break
                }

                val entry = Json5Value(reader)
                entries.add(entry)
                values.add(entry)

                while (!reader.exhausted()) {
                    val value = reader.parseOrNull(
                        Json5Whitespace::invoke,
                        Json5Comment::invoke
                    )
                    if (value == null) break
                    entries.add(value)
                }
                if (reader.peek() == ',') {
                    entries.add(reader.take()!!)
                } else {
                    break
                }
            }
            reader.expect(']')
            return ArrayContent(entries, values)
        }

    }

    class ArrayContent(
        val data: List<Any>,
        val values: List<Json5Value>
    ) : Content() {

        override val entries: Iterable<*>
            get() = iterable {
                yield('[')
                yieldAll(data)
                yield(']')
            }

    }

}